error_reporting.py

```
@@ -0,0 +1,479 @@
+            new_var = ap.convert_expr_dsa_vars_to_ap(upd.var)
+            new_expr = ap.convert_expr_dsa_vars_to_ap(upd.expr)
+            return source.Update(var=new_var, expr=new_expr)
+        upds = tuple(map(convert_update, node.upds))
+        return source.NodeBasic(origin=node.origin, upds=upds, succ=node.succ)
+
+    elif isinstance(node, source.NodeAssume):
+        new_expr = ap.convert_expr_dsa_vars_to_ap(node.expr)
+        return source.NodeAssume(origin=node.origin, succ=node.succ, expr=new_expr)
+    elif isinstance(node, source.NodeAssert):
+        new_expr = ap.convert_expr_dsa_vars_to_ap(node.expr)
+        return source.NodeAssert(origin=node.origin, succ=node.succ, expr=new_expr)
+    elif isinstance(node, source.NodeCall):
+        args = tuple(map(ap.convert_expr_dsa_vars_to_ap, node.args))
+        rets = tuple(map(ap.convert_expr_dsa_vars_to_ap, node.rets))
+        return source.NodeCall(origin=node.origin, succ=node.succ, args=args, rets=rets, fname=node.fname)
+    elif isinstance(node, source.NodeCond):
+        new_expr = ap.convert_expr_dsa_vars_to_ap(node.expr)
+        return source.NodeCond(origin=node.origin, expr=new_expr, succ_then=node.succ_then, succ_else=node.succ_else)
+    else:
+        assert_never(node)
+
+
+def debug_func_smt(func: dsa.Function) -> Tuple[FailureReason, source.NodeName, Optional[source.NodeName]]:
+    prog = ap.make_prog(func)
+    q: set[source.NodeName] = set([func.cfg.entry])
+    not_taken_path: set[source.NodeName] = set([])
+    visited: set[source.NodeName] = set([])
+    while len(q) != 0:
+        node_name = q.pop()
+        visited = visited.union(set([node_name]))
+        node = func.nodes[node_name]
+        not_taken_path_and_node = not_taken_path.union(set([node_name]))
+        node_smtlib = smt.make_smtlib(prog, not_taken_path_and_node)
+        consistent, node_sat = get_sat(node_smtlib)
+        assert consistent
+        # we do not care about the Err and Ret node
```

I see why we should ignore Err, but not why we should ignore Ret